### PR TITLE
Partitioned write into s3 table function

### LINKED ISSUE #15171: Partitioned write into s3 table function / S3 table
Allow to specify PARTITION BY expression for S3 table engine.
When writing to this table, multiple files will be created, with names prefixed (?) by partition key.

`PARTITION BY rand() % N` will be also applicable for creation of multiple files instead of one.

Motivation:
- more efficient to read;
- easier manipulation by other data processing systems.

Details of implementation:
- if s3 path contains wildcard (`*`), we will substitute this wildcard by the value of partition key; it can be located in file name or in part of path (in the latter case multiple directories will be created).

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Partitioned write into s3 table function

Detailed description / Documentation draft:

...

Shall resolve #15171

### CODE REVIEW

**Comment by excitoon** [comment]
Failed tests are unrelated.

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:411]:
```suggestion
    {
        /// writeSuffix for each child writer called in getWriterForPartition on writer creation
```

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:380]: Maybe we need to sanitize partition key in some way?

For example following code inserting string containing zero bytes

<details>
<summary>Python code snippet</summary>

```
from clickhouse_driver import Client
ch_client = Client('localhost', 9000)

command = """
INSERT INTO TABLE FUNCTION 
s3('http://localhost:9001/foo/test_{_partition_id}.csv', 
    'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') 
PARTITION BY val
VALUES
"""

ch_client.execute(command, [
    {'id': 1, 'val': '\x00\x00\x00\x00'},
])
```

</details>

Raises

<details>
<summary>ServerException</summary>

```
ServerException: Code: 499.
DB::Exception: The request signature we calculated does not match the signature you provided. Check your key and signing method.. Stack trace:

0. /build/build_docker/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x11fada0c in /home/vdimir/_tmp/clickhouse-23051
1. /build/build_docker/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0x8f2ac9a in /home/vdimir/_tmp/clickhouse-23051
2. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:233: DB::WriteBufferFromS3::makeSinglepartUpload() @ 0xf03d7e8 in /home/vdimir/_tmp/clickhouse-23051
3. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:112: DB::WriteBufferFromS3::finalizeImpl() @ 0xf03ce0b in /home/vdimir/_tmp/clickhouse-23051
4. /build/build_docker/../src/IO/WriteBufferFromS3.cpp:0: DB::WriteBufferFromS3::finalize() @ 0xf03cd61 in /home/vdimir/_tmp/clickhouse-23051
5. /build/build_docker/../src/Storages/StorageS3.cpp:329: DB::StorageS3BlockOutputStream::writeSuffix() @ 0xf944b28 in /home/vdimir/_tmp/clickhouse-23051
6. /build/build_docker/../src/Storages/StorageS3.cpp:0: DB::StorageS3PartitionedBlockOutputStream::writeSuffix() @ 0xf942c3a in /home/vdimir/_tmp/clickhouse-23051
7. /build/build_docker/../contrib/libcxx/include/exception:176: DB::PushingToViewsBlockOutputStream::writeSuffix() @ 0xf357143 in /home/vdimir/_tmp/clickhouse-23051
8. /build/build_docker/../src/Server/TCPHandler.cpp:581: DB::TCPHandler::processInsertQuery(DB::Settings const&) @ 0xfccf69f in /home/vdimir/_tmp/clickhouse-23051
9. /build/build_docker/../contrib/libcxx/include/functional:2236: DB::TCPHandler::runImpl() @ 0xfcca3cd in /home/vdimir/_tmp/clickhouse-23051
10. /build/build_docker/../src/Server/TCPHandler.cpp:1623: DB::TCPHandler::run() @ 0xfcd5c19 in /home/vdimir/_tmp/clickhouse-23051
11. /build/build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x11f215a7 in /home/vdimir/_tmp/clickhouse-23051
12. /build/build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114: Poco::Net::TCPServerDispatcher::run() @ 0x11f21a3c in /home/vdimir/_tmp/clickhouse-23051
13. /build/build_docker/../contrib/poco/Foundation/include/Poco/ScopedLock.h:36: Poco::PooledThread::run() @ 0x120088d9 in /home/vdimir/_tmp/clickhouse-23051
14. /build/build_docker/../contrib/poco/Foundation/include/Poco/SharedPtr.h:156: Poco::ThreadImpl::runnableEntry(void*) @ 0x120065e6 in /home/vdimir/_tmp/clickhouse-23051
15. start_thread @ 0x76db in /lib/x86_64-linux-gnu/libpthread-2.27.so
16. /build/glibc-OTsEL5/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97: __clone @ 0x12188f in /usr/lib/debug/lib/x86_64-linux-gnu/libc-2.27.so
```
</details>

It's not a big issue, but I'm not sure that something more dangerous can't happen. Any ideas?

**Review by vdimir — COMMENTED** [review]
Thank you for you changes!

I took a first look and left comments.

Also maybe I'll dig into details once again.

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:380]: There is no strong restrictions on key names in S3 :( And no code for validation in SDK. As for `\x00` symbol, it is not technically forbidden, probably this can be even fixed.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

Comment (vdimir) [reply src/Storages/StorageS3.cpp:382]: It's stated that `You can use any UTF-8 character in an object key name`, but we can generate invalid UTF-8 string. Anyway, there's `Safe characters`, do you mind to add validation according to this, throwing exception if not valid string found?

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:58]: If string already "escaped" with `{}`, do we nedd `_` at the beginning. Looks like overkill, maybe just `{partition_id}`. But actually I'm not sure which is better.

**Comment by excitoon** [comment]
I did that to match existing virtual column name, additionally saving
future possibility to implement expressions or formats (may be).

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:382]: We already have people who use unsafe characters like `+` (specifically fixed that) and also there may be other non AWS S3 implementations🤷🏼‍♂️🤷🏼‍♂️

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:58]: I did that to match existing virtual column name, additionally saving
future possibility to implement expressions or formats (may be).

File: src/Parsers/ParserInsertQuery.cpp
Comment (vdimir) [reply src/Parsers/ParserInsertQuery.cpp:70]: Maybe let's add test for it? I want to push it to PR, but seems that there's no permission. https://github.com/vdimir/ClickHouse/commit/70e87b3f46ccba6193345f707785e72bd285b6e7

File: src/Common/ErrorCodes.cpp
Comment (vdimir) [inline src/Common/ErrorCodes.cpp:564]:
```suggestion
    M(594, NOT_IMPLEMENTED) \
```

File: src/Interpreters/InterpreterInsertQuery.cpp
Comment (vdimir) [inline src/Interpreters/InterpreterInsertQuery.cpp:43]:
```suggestion
    extern const int NOT_IMPLEMENTED;
```

File: src/Storages/StorageS3.cpp
Comment (vdimir) [reply src/Storages/StorageS3.cpp:382]: But my example above return error and message is not clear what happened

**Review by vdimir — COMMENTED** [review]
Ok, generally it's good, except non-validated keys and unclear error message around it.

File: src/Common/ErrorCodes.cpp
Comment (vdimir) [reply src/Common/ErrorCodes.cpp:564]: And I also cannot commit this via web interface. You can apply three suggestion or cherry pick https://github.com/ClickHouse/ClickHouse/commit/9c1cb601e957f96e5bbceb811bb0070060a63d5a

Comment (excitoon) [reply src/Common/ErrorCodes.cpp]: Pushed existing `NOT_IMPLEMENTED`.

File: src/Parsers/ParserInsertQuery.cpp
Comment (excitoon) [reply src/Parsers/ParserInsertQuery.cpp:70]: Added commit and permission.

**Review by vdimir — APPROVED** [review]

**Comment by vdimir** [comment]
I've added partition key validation. We can change it later if it'll be too strict, I suppose it's better than adding restriction and broke some behavior in future.

**Comment by excitoon** [comment]
For some reason GitHub does not show your commits separately https://github.com/ClickHouse/ClickHouse/pull/23051/files/82afef5..e91988b

File: src/Storages/StorageS3.cpp
Comment (excitoon) [inline src/Storages/StorageS3.cpp:490]: I guess, that this is not very good, alphabetic characters from other languages are disabled now.

Comment (excitoon) [reply src/Storages/StorageS3.cpp:490]: I would also add links `https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html` and `https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html` to the comments in this case.

File: src/Storages/StorageS3.cpp
Comment (excitoon) [inline src/Storages/StorageS3.cpp:478]: Why partition key can not be empty? Hypothetically, it can be a "sorted comma-separated list of binary features".

File: src/Storages/StorageS3.cpp
Comment (excitoon) [inline src/Storages/StorageS3.cpp:475]: Minor typo — `validatePar`**`t`**`itionKey`.

File: src/Storages/StorageS3.cpp
Comment (vdimir) [reply src/Storages/StorageS3.cpp:478]: If we allow empty keys such case produce unclear error:


`insert into table function s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val values (1, '');`

```
Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:477]: Also we have to check maximal length

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:478]: Same in this case:
```
cristal :) insert into table function s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') values (1, '');

INSERT INTO FUNCTION s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') VALUES

Query id: f4c92770-b6fc-45f2-a6a6-5e8af62126b4


1 rows in set. Elapsed: 0.037 sec. 

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

I think we shall remove any validation in this pull request, BUT implement it in another, in `WriteBufferFromS3`/`ReadBufferFromS3`.

**Comment by excitoon** [comment]
@vdimir Test is fixed but there is something wrong with RabbitMQ in integration tests.

**Comment by excitoon** [comment]
@vdimir Can you look through it again, I made some rework after breaking changes.

**Comment by vdimir** [comment]
> Test is fixed but there is something wrong with RabbitMQ in integration tests

RabbitMQ tests broken in master

> Can you look through it again? I made some rework after breaking changes

Ok, I'm going to look at the changes

File: src/Storages/StorageS3.cpp
Comment (vdimir) [reply src/Storages/StorageS3.cpp:483]: But actually simple loop and checking `*pos < 32 && *pos != '{' && *pos != '}' && ...` or something like this will be better, because in current implementation we perform 3 iteration over whole string on "happy path"

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:470]: Should we check that result bucket name and key are not empty? And also maximal length?

**Review by vdimir — APPROVED** [review]
Ok, only couple of comments

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:470]: @vdimir Bucket name is being checked in `S3::URI`, and empty key is allowed in AWS.

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:476]: +

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:484]:
```
vladimir@cristal:~/find_first$ cat 1.cpp 
int main(int argc, char ** argv)
{
    bool result;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;
        for (char * i = argv[0]; *i; ++i)
        {
            if (*i < 32 || *i == '{' || *i == '}' || *i == '*' || *i == '?')
            {
                result = false;
                break;
            }
        }
    }
    return result ? 0 : 1;
}

```

```
#include "/home/vladimir/ClickHouse9/base/common/find_symbols.h"

int main(int argc, char ** argv)
{
    bool result;
    size_t len = 0;
    for (; argv[0][len] == 0; ++len);
    char *end = argv[0]+len;
    for (int j = 0; j < 100000000; ++j)
    {
        result = true;

        const char * first_symbol = find_first_symbols<
                '\x00', '\x01', '\x02', '\x03', '\x04', '\x05', '\x06', '\x07',
                '\x08', '\x09', '\x0A', '\x0B', '\x0C', '\x0D', '\x0E', '\x0F'>(argv[0], end);
        if (first_symbol == end)
        {
            first_symbol = find_first_symbols<
                    '\x10', '\x11', '\x12', '\x13', '\x14', '\x15', '\x16', '\x17',
                    '\x18', '\x19', '\x1A', '\x1B', '\x1C', '\x1D', '\x1E', '\x1F'
                >(argv[0], end);
            if (first_symbol == end)
            {
                first_symbol = find_first_symbols<'{', '}', '*', '?'>(argv[0], end);
                if (first_symbol == end)
                {
                    result = false;
                }
            }
        }
    }
    return result ? 0 : 1;
}
```

```
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 1.cpp -o 1
vladimir@cristal:~/find_first$ g++ -std=c++17 -O2 2.cpp -o 2
vladimir@cristal:~/find_first$ time ./1 abcdefghjkja

real	0m0.378s
user	0m0.367s
sys	0m0.000s
vladimir@cristal:~/find_first$ time ./2 abcdefghjkja

real	0m0.115s
user	0m0.114s
sys	0m0.000s
```

**Comment by excitoon** [comment]
@vdimir Looks like it fixed everything. What are next steps?

File: src/Storages/StorageS3.cpp
Comment (excitoon) [reply src/Storages/StorageS3.cpp:484]: Okay

File: src/Storages/StorageS3.cpp
Comment (vdimir) [reply src/Storages/StorageS3.cpp:470]: Do you mean that it's checked on server-side?

```
INSERT INTO FUNCTION s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```


```
INSERT INTO FUNCTION s3('http://localhost:9001/{_partition_id}/key.csv', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val VALUES

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. (S3_ERROR)
```

File: src/Storages/StorageS3.cpp
Comment (vdimir) [inline src/Storages/StorageS3.cpp:490]: Strange error `fmt::v7::format_error: cannot switch from manual to automatic argument indexing.`, maybe `StringRef` not supported here
```suggestion
                throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT,
                        "Illegal character '\\x{0:02x}' in partition key",
                        invalid_char_byte);
```

Comment (excitoon) [reply src/Storages/StorageS3.cpp]: This is not because of `StringRef`.

Comment (excitoon) [reply src/Storages/StorageS3.cpp]: Actually, new code does not work at all.

**Comment by vdimir** [comment]
> @vdimir Looks like it fixed everything. What are next steps?

@excitoon Ok, I'll merge when CI passed. Also look at this comment about checking length https://github.com/ClickHouse/ClickHouse/pull/23051#discussion_r678997963

**Comment by excitoon** [comment]
@vdimir Fixed it, but there are few failing unrelated tests.

File: src/IO/S3Common.cpp
Comment (vdimir) [inline src/IO/S3Common.cpp:699]: Need to be removed?

Comment (excitoon) [reply src/IO/S3Common.cpp:699]: I would leave it there as FIXME/documentation. I guess we may implement more strict checks at some point.

Comment (vdimir) [reply src/IO/S3Common.cpp:699]: But now we wont check that input is UTF8?

Comment (excitoon) [reply src/IO/S3Common.cpp:699]: Yes, because in COS non-UTF-8 is allowed. I guess in separate PR we may implement different checks depending on type of backend

Comment (vdimir) [reply src/IO/S3Common.cpp:699]: So, do we want to leave it without validation? I remember that we decided to validate utf8 + no ascii control chars + no glob chars + length. All is checked except utf8.

Comment (excitoon) [reply src/IO/S3Common.cpp:699]: Yes, because otherwise COS would be broken 🤷‍♂️

**Comment by excitoon** [comment]
I made a rebase in order to pass integration tests.

**Comment by excitoon** [comment]
@vdimir What are our next steps?

**Comment by vdimir** [comment]
> @vdimir What are our next steps?

I suppose we should add validation as discussed earlier. I added it and rebased to master, let's wait for CI and merge.

**Comment by excitoon** [comment]
@vdimir We can not add validation, we would break `COS` table function by these changes. :man_shrugging:

**Comment by vdimir** [comment]
> @vdimir We can not add validation, we would break `COS` table function by these changes. 🤷‍♂️

Validation would be performed only for new feature.

**Comment by excitoon** [comment]
@vdimir What is the logic behind that this feature will not support COS fully, though we can easily achieve that?

**Comment by vdimir** [comment]
Commit https://github.com/ClickHouse/ClickHouse/pull/23051/commits/98acccbdcc39ec95a22626b7b256489474338dc2 is only merged master to resolve conflicts, no changes, build in Fast test passed, will merge. Full CI passed on commit https://github.com/ClickHouse/ClickHouse/pull/23051/commits/6fe63a80bc6addbb063074f665f71f2a7842612c

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-13742

**Comment by alexey-milovidov** [comment]
@excitoon @vdimir an issue has been found: #34348

**Comment by alexey-milovidov** [comment]
@excitoon @vdimir a bug has been found: #38199

**Pull request merged**
